docs/advanced-features/middleware.md

```diff
@@ -1,102 +1,361 @@
-- [Redirects and rewrites](https://github.com/vercel/examples/tree/main/edge-functions)
-- [Handling unsupported browsers](https://github.com/vercel/examples/tree/main/edge-functions)
-- [Feature flags and A/B tests](https://github.com/vercel/examples/tree/main/edge-functions)
-- [Advanced i18n routing requirements](https://github.com/vercel/examples/tree/main/edge-functions)
+### Why does `redirect` use 307 and 308?
 
-## Execution Order
+When using `redirect()` you may notice that the status codes used are `307` for a temporary redirect, and `308` for a permanent redirect. While traditionally a `302` was used for a temporary redirect, and a `301` for a permanent redirect, many browsers changed the request method of the redirect, from a `POST` to `GET` request when using a `302`, regardless of the origins request method.
 
-Middleware runs directly after `redirects` and `headers`, before the first filesystem lookup. This excludes `/_next` files.
+Taking the following example of a redirect from `/users` to `/people`, if you make a `POST` request to `/users` to create a new user, and are conforming to a `302` temporary redirect, the request method will be changed from a `POST` to a `GET` request. This doesn't make sense, as to create a new user, you should be making a `POST` request to `/people`, and not a `GET` request.
 
-## Deployment
+The introduction of the `307` status code means that the request method is preserved as `POST`.
 
-Middleware uses a [strict runtime](/docs/api-reference/edge-runtime.md) that supports standard Web APIs like `fetch`. This works out of the box using `next start`, as well as on Edge platforms like Vercel, which use [Edge Functions](http://www.vercel.com/edge).
+- `302` - Temporary redirect, will change the request method from `POST` to `GET`
+- `307` - Temporary redirect, will preserve the request method as `POST`
 
-## Related
+The `redirect()` method uses a `307` by default, instead of a `302` temporary redirect, meaning your requests will _always_ be preserved as `POST` requests.
 
-<div class="card">
-  <a href="/docs/api-reference/next/server.md">
-    <b>Middleware API Reference</b>
-    <small>Learn more about the supported APIs for Middleware.</small>
-  </a>
-</div>
+If you want to cause a `GET` response to a `POST` request, use `303`.
+
+[Learn more](https://developer.mozilla.org/en-US/docs/Web/HTTP/Redirections) about HTTP Redirects.
+
+### How do I access Environment Variables?
+
+`process.env` can be used to access [Environment Variables](/docs/basic-features/environment-variables.md) from Middleware. These are evaluated at build time, so only environment variables _actually_ used will be included.
+
+Any variables in `process.env` must be accessed directly, and **cannot** be destructured:
+
+```ts
+// Accessed directly, and not destructured works. process.env.NODE_ENV is `"development"` or `"production"`
+console.log(process.env.NODE_ENV)
+// This will not work
+const { NODE_ENV } = process.env
```

Comment:
Confirmed https://github.com/vercel/runtimes/blob/main/packages/edge-functions-bridge/src/extract-env-vars.ts#L63

Reply:
@cramforce this function does not work on Next.js though, as Next.js declares its own env vars because it analyzes the code pre-bundling:
https://github.com/vercel/next.js/blob/977ff5247222acfdf0b7429fa08c9f95adcb030e/packages/next/build/webpack/plugins/middleware-plugin.ts#L172-L176

Reply:
I think that destructuring works but we need to check it with tests 💡

Reply:
Uhm, can we fix?

Reply:
![👍 ](https://user-images.githubusercontent.com/2054772/172641921-d30e8c12-e0ba-4fa5-a154-95aa3f6148aa.png)

... or at least I will try to!